### PR TITLE
[#17158][fix] Reject NaN top_p, min_p and temperature in SamplingParams

### LINKED ISSUE #17158: [Bug]: SamplingParams accepts NaN for top_p, min_p and temperature
### System Info

- CPU architecture: N/A (reproducible with a pure-Python snippet, no GPU required)
- GPU: N/A
- TensorRT-LLM branch: `main`
- TensorRT-LLM commit: `a9544e0ded815ba487e1ee3e670890905b93eff0`
- OS: N/A

The report concerns `SamplingParams._validate()` in `tensorrt_llm/sampling_params.py`.

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

`SamplingParams._validate()` bounds `top_p`, `min_p` and `temperature` with positive
range checks (`tensorrt_llm/sampling_params.py:375-382`):

```python
if self.top_p is not None and (self.top_p < 0 or self.top_p > 1):
    raise ValueError(f"require 0 <= top_p <= 1, got top_p={self.top_p}")
...
if self.min_p is not None and (self.min_p < 0 or self.min_p > 1):
    raise ValueError(f"require 0 <= min_p <= 1, got min_p={self.min_p}")
if self.temperature is not None and self.temperature < 0:
    raise ValueError(f"require temperature >= 0, got temperature={self.temperature}")
```

Every comparison against NaN evaluates to `False`, so NaN passes all three checks:

```python
from tensorrt_llm.sampling_params import SamplingParams

nan = float("nan")
print(SamplingParams(top_p=nan).top_p)              # nan
print(SamplingParams(min_p=nan).min_p)              # nan
print(SamplingParams(temperature=nan).temperature)  # nan
```

The two checks immediately below, added for the top-p decay parameters, are written as
negated range checks and therefore already reject NaN
(`tensorrt_llm/sampling_params.py:389-392`):

```python
if self.top_p_decay is not None and not 0.0 < self.top_p_decay <= 1.0:
    raise ValueError(...)
```

```python
SamplingParams(top_p_decay=nan)
# ValueError: require 0 < top_p_decay <= 1, got top_p_decay=nan
```

So the file is inconsistent with itself: some sampling parameters reject NaN, the three
most commonly used ones do not.

### Expected behavior

`SamplingParams` rejects NaN for `top_p`, `min_p` and `temperature` at construction
time, with the same error messages already used for out-of-range values.

### actual behavior

NaN is accepted and forwarded to the sampling backend. There is no downstream guard on
these values: `_torch/pyexecutor/sampler/sampler_strategy.py:325-337` divides the logits
by the raw temperature tensor, and `sampler/ops/flashinfer.py:216` passes it straight to
the fused softmax. The result is a request that samples from an undefined distribution
instead of failing with a clear message at the API boundary.

### additional notes

Only NaN is in scope here. `temperature=float("inf")` is also accepted today; it is
well-defined (it flattens the distribution) rather than broken, so this report does not
propose changing it.

Related but distinct: #15715 asks for very small non-zero temperatures to be clamped for
numerical stability. That is a behavior change to valid input; this report is about
input that no range check was ever meant to admit.

I have a fix and will open a PR referencing this issue.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and checked the [documentation](https://nvidia.github.io/TensorRT-LLM/) and [examples](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples) for answers to frequently asked questions.

### PR DESCRIPTION
**Pull request by Yigtwxx**
## Description

Fixes #17158.

`SamplingParams._validate()` bounded `top_p`, `min_p` and `temperature` with positive
range checks of the form `value &lt; low or value &gt; high`. Every comparison against NaN
evaluates to `False`, so `SamplingParams(top_p=float(&#34;nan&#34;))` and its `min_p` /
`temperature` equivalents passed validation unchanged and were forwarded to the sampling
backend, which applies no guard of its own — `sampler_strategy.py:325-337` divides the
logits by the raw temperature tensor and `ops/flashinfer.py:216` hands it to the fused
softmax. The user got a request sampling from an undefined distribution rather than a
clear error at the API boundary.

The three checks are rewritten as negated range checks (`not 0 &lt;= top_p &lt;= 1`,
`not 0 &lt;= min_p &lt;= 1`, `not temperature &gt;= 0`). This rejects NaN and leaves the accepted
set otherwise identical, with the existing error messages unchanged. It is also the form
the `top_p_decay` and `top_p_min` checks a few lines below already use — those two
consequently reject NaN today, which is what made the file inconsistent with itself.

Deliberately out of scope: `temperature=float(&#34;inf&#34;)` is still accepted. It flattens the
distribution rather than corrupting it, so rejecting it would be a behavior change to
input that currently works. #15715, which asks for very small non-zero temperatures to be
clamped, is likewise a separate discussion about valid input and is not addressed here.

No API signature changes; `tests/unittest/api_stability` is unaffected.

## Test Coverage

`tests/unittest/llmapi/test_sampling_params.py` (CPU-only, no model weights):

- `test_sampling_params_rejects_nan` — parametrized over `top_p`, `min_p`, `temperature`;
  each must raise `ValueError`. This is the regression guard: all three cases fail
  against `main`, where NaN is accepted, and pass with this change.
- `test_sampling_params_rejects_out_of_range` — the previously covered rejections
  (`-0.1`, `1.1`, `temperature=-1.0`) still raise, confirming the rewrite did not narrow
  the checks.
- `test_sampling_params_accepts_in_range_values` — boundary and typical values (`0.0`,
  `0.9`, `1.0`, `0.5`, `temperature=0.0`, `temperature=1.0`) are still accepted and round
  trip unchanged.

Results: 16 cases, all passing on this branch. Against `main` with only the test changes
applied, exactly the 3 NaN cases fail and the other 13 pass, so the accepted set is
otherwise unchanged.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit&#39;s summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

## Dev Engineer Review

- `SamplingParams._validate()` now rejects `NaN` for `top_p`, `min_p`, and `temperature`.
- Existing error messages and valid range behavior remain unchanged.
- `top_k` validation is unchanged.
- No public API changes are introduced.
- The test-list entry uses the correct test path and does not expand test scope beyond the CPU-only unit test.

**Verdict: sufficient**

## QA Engineer Review

- Added coverage for:
  - `NaN` rejection for `top_p`, `min_p`, and `temperature`.
  - Existing out-of-range rejection.
  - Accepted boundary values.
  - Accepted representative in-range values.
- The test file is registered in `tests/integration/test_lists/test-db/l0_cpu.yml`.
- CI coverage is sufficient for the added regression tests.

**Verdict: sufficient**

## Summary

- Rejects `NaN` for `top_p`, `min_p`, and `temperature` in `SamplingParams._validate()`.
- Preserves existing validation errors and behavior for valid and previously rejected values.
- Adds validation tests and registers them in the CPU test suite.
- No public API changes.
- Positive-infinite `temperature` and small-temperature clamping remain out of scope.

## Dev Engineer Review

- The negated range checks correctly reject `NaN` because comparisons with `NaN` evaluate to false.
- `top_k` validation remains unchanged.
- The change preserves valid boundary values and existing out-of-range behavior.
- The CPU test registration uses the correct test path and has no unintended scope change.
- No performance or API consistency concerns were identified.
- Local tests, test-list AST validation, and pre-commit checks passed.
- CI failed, but reported failures appeared related to shared pipeline or infrastructure issues.

## QA Engineer Review

- Added coverage for:
  - `NaN` rejection for `top_p`, `min_p`, and `temperature`.
  - Out-of-range values.
  - Boundary and representative valid values.
- The test file is registered in `tests/integration/test_lists/test-db/l0_cpu.yml`.
- The test coverage is included in the CPU CI suite.
- Verdict: sufficient.

### CODE REVIEW

**Review by BowenFu — APPROVED** [review]
Verified the rewrite is exactly equivalent to the old checks for every non-NaN input, including the 0/1 boundaries and ±inf, so the only behavior change is that NaN now raises. `test_sampling_params_accepts_in_range_values` is the right guard against over-tightening.

Not blocking: `tests/unittest/llmapi/test_sampling_params.py` isn't in any `test-db/` list, so these tests won't actually run in pre-merge L0. Worth registering.

File: tests/unittest/llmapi/test_sampling_params.py [hunk 1]
```diff
@@ -115,6 +115,44 @@ def test_completion_logprobs_assignment_revalidates():
         request.to_sampling_params(backend="pytorch")
 
 
+@pytest.mark.parametrize("field", ["top_p", "min_p", "temperature"])
+def test_sampling_params_rejects_nan(field):
+    with pytest.raises(ValueError, match=field):
+        SamplingParams(**{field: float("nan")})
```
[the remark this answers is not in the document]
Comment (Yigtwxx) [reply tests/unittest/llmapi/test_sampling_params.py:123]: Good catch — confirmed and fixed in 5fdb20af51.

`tests/unittest/llmapi/test_sampling_params.py` predates this PR but was never registered in any list, so none of its cases ran in CI, including the NaN regression tests added here. It is now listed in the `# llmapi` group of the A10 pre-merge stage (`tests/integration/test_lists/test-db/l0_a10.yml`), next to `test_reasoning_parser.py`.

I did not add it to the QA lists. Those cover GPU/model-backed functional tests; the CPU-only llmapi unit modules (`test_llm_args.py`, `test_reasoning_parser.py`, `test_utils.py`, ...) are all registered in `test-db` only, and this module follows the same shape. Happy to add a QA entry if a reviewer prefers otherwise.

One note for reviewers: because the module never ran in CI before, this run is the first for its pre-existing cases as well, not just the three added here.

**Review by xinhe-nv — APPROVED** [review]

**Review by StanleySun639 — APPROVED** [review]

**Comment by Yigtwxx** [comment]
Could a maintainer trigger CI on this one with `/bot run`?

I posted `/bot run` yesterday but it didn't start a pipeline — I don't have write access on this repo, so the bot appears to ignore the command from me. The head commit `5fdb20a` currently has DCO as its only check and no Jenkins status, which is what keeps the PR in a blocked state.

Everything else looks settled: the review feedback about registering the test module in the test lists was addressed in `5fdb20a` (added to the `# llmapi` group of `tests/integration/test_lists/test-db/l0_a10.yml`) and confirmed resolved, and the PR has two approvals. Nothing is pending on my side.

The change is CPU-only and touches `SamplingParams._validate()` plus one test-list entry, so a standard `/bot run` should be sufficient.

**Comment by Yigtwxx** [comment]
Thanks for triggering CI @zhaoyangwang-nvidia. I can't open the failure report or the Jenkins job (both are on the internal network), so could you share which stage/test `L0_MergeRequest_PR` pipeline #51648 failed on?

Meanwhile, some evidence that suggests the failure isn't specific to this PR: in the 2026-08-03 → 2026-08-05 window, **13 of 13 completed `L0_MergeRequest_PR` pipelines failed** across 5 unrelated PRs from 5 different authors (#17107, #17165, #17213, #17230, #16992), covering unrelated areas — flashinfer attention autotuner, Jenkins groovy, EXAONE config, Mamba cache + disagg transfer, CuTe DSL kernels. #17107 re-triggered the same commit (`30fdacf`) three times (#51616, #51669, #51713) and all three failed. The oldest red run I found is from 2026-07-29, so this looks like it predates my branch.

One thing worth noting for anyone skimming: in the `tensorrt-cicd` comments, `PR_Github #NNNNN [ run ] completed with state SUCCESS` refers to the outer helper job, not the pipeline — in 7 of those 13 cases the helper said `SUCCESS` while the `L0_MergeRequest_PR pipeline` line in the same comment said `'FAILURE'`.

On this PR's side, the change is CPU-only: three comparisons in `SamplingParams._validate()` plus one test-list entry. I ran the new tests locally (17/17 pass), covering NaN rejection, out-of-range rejection, in-range acceptance, and that `temperature=inf` is still accepted. I also grepped `tensorrt_llm/` and `tests/` and found no call site that passes NaN to `top_p`/`min_p`/`temperature`, so the stricter check can only reject values that previously slipped through silently.

The branch is currently 58 commits behind `main`. Happy to rebase onto latest `main` and have CI re-triggered if that helps — just let me know.

**Comment by Yigtwxx** [comment]
Rebased onto current `main` (`fd0b4bb`) — the branch had gone into a conflicting state. New head is `30897c3`.

The conflict was only in the test-list file, and resolving it turned out to be a real change rather than a mechanical one: `main` has since split the CPU-only llmapi unit tests out of `l0_a10.yml` into the new `l0_cpu.yml` (`test_llm_utils`, `test_gc_utils`, `test_reasoning_parser`, `test_serialization`, `test_utils`, `test_kv_cache_dtype_override`, `test_request_priority` all moved). Since `test_sampling_params.py` is CPU-only — the tests just construct `SamplingParams` and assert on validation — **the registration now goes in `l0_cpu.yml`** instead, inserted alphabetically between `test_request_priority.py` and `test_serialization.py`. `l0_a10.yml` is left exactly as `main` has it. I updated the PR description accordingly.

`sampling_params.py` and the test file rebased cleanly; the fix itself is unchanged, and the comment it adds still holds — `top_p_decay` / `top_p_min` on current `main` still use the negated form.

Verified after the rebase: the new tests still pass locally (16 cases in the file), and `pre-commit` is clean on the diff, including the `Validate test list entries exist in source files (AST)` hook that checks the new `l0_cpu.yml` entry resolves.

Ready for `/bot run` whenever someone can trigger it. My earlier question still stands if the next run also fails — I can't see which stage failed, so I'd need that from whoever triggers it.

**Comment by Yigtwxx** [comment]
@coderabbitai Re: the outside-diff finding on `tests/integration/test_lists/test-db/l0_cpu.yml` — not applying it. Two reasons:

1. **It is not a new file.** GitHub reports it as `status: modified`, `+1/-0`; this PR's commit is `index 3bb62ffd3b..5a1a08f2e1`, a single added line (`unittest/llmapi/test_sampling_params.py`) in the existing `# llmapi` group. The file existed before this branch. `l0_cpu_x86.yml` and `l0_cpu_arm.yml` were consolidated into `l0_cpu.yml` upstream, not here — that is likely what made it look newly supplied.

2. **The header would break the convention, not restore it.** 50 of the 51 files under `tests/integration/test_lists/test-db/` start directly with `version: 0.0.1` and carry no copyright header (`l0_b200_perf_sanity.yml` is the lone exception). There is also no `insert-license` hook in `.pre-commit-config.yaml` covering YAML. Adding a header to this one file as a side effect of a one-line test registration would be an unrelated, inconsistent change and out of scope for this PR.

Happy to be corrected if a maintainer wants headers rolled out across `test-db/` — but that belongs in its own PR, not this one.

**Comment by Yigtwxx** [comment]
@zhaoyangwang-nvidia — quick ping since you triggered the last run: the branch has been rebased onto current `main` (`fd0b4bb`), so the head is now `30897c3` and the earlier #63699 result on `5fdb20a` no longer applies. `30897c3` has DCO as its only check.

Could you re-trigger `/bot run` when you get a chance? My own `/bot run` comments don't start a pipeline — I don't have write access here, so the bot ignores them.

And if this run also comes back red, could you paste which stage/test failed? The report and Jenkins links are internal-only, so I can't see them from outside.

No other blockers on my side: four approvals, review feedback addressed, and the CodeRabbit copyright-header finding answered [above](https://github.com/NVIDIA/TensorRT-LLM/pull/17159#issuecomment-5180478745).

**Comment by Yigtwxx** [comment]
Correction to my earlier comment (https://github.com/NVIDIA/TensorRT-LLM/pull/17159#issuecomment-5180231430): the "13 of 13 completed pipelines failed" figure in it is wrong. Retracting it, since I'd rather it not feed into anyone's triage.

I redid the count over a much larger sample — 70 PRs, every `tensorrt-cicd` comment between 2026-08-03 00:00Z and 2026-08-04 15:30Z — reading the inner `L0_MergeRequest_PR pipeline #N completed with status: '...'` line rather than the outer `PR_Github #N [ run ] completed with state ...` line (those two disagree, 25 times in this sample), and excluding `[ skip ]` invocations, which report success without running any tests:

| | |
|---|---|
| Pipelines reporting a status | 129 |
| SUCCESS | 24 (18.6%) |
| FAILURE | 102 |
| UNSTABLE | 3 |
| PRs that ran a pipeline | 45 |
| — with at least one green run | 24 |
| — never green | 21 |

So `L0_MergeRequest_PR` is red for most runs in that window, but green is plainly reachable, and four of the five PRs I cited (#17107, #17165, #17213, #17230) did have green runs in the same window — #17213 in pipelines 63480 and 63757, for instance. My original number came from an ~8-PR sample, which is biased in an obvious way once pointed out: a failing PR gets re-run repeatedly and keeps appearing, while a passing one merges and drops out.

That removes the "this isn't specific to my PR" inference, so the ask stands on its own merits rather than on that claim:

@zhaoyangwang-nvidia — could you re-trigger `/bot run` when you have a moment? The branch was rebased onto `fd0b4bb`, so the head is now `30897c3` and the #63699 result on `5fdb20a` no longer applies. And if the new run comes back red, could you paste which stage/test failed? The CI report and Jenkins links are internal-only, so I can't read them from outside.
